refactor(relation): table relation_unit refers relation_endpoint instead of relation

gfouillet opened the pull request:
This change introduce a direct reference from a relation unit to the application it belongs, instead of requiring a JOIN through unit<application_uuid>application_endpoint.

This simplifies the request GetRelationApplicationEndpointName, but make the insert a bit more complicated (insertion is done through relationUnitUUID) as long as few other requests.

However, new JOIN are unidirectional, which make the overall reasoning about schema simpler.

- `domain/relation/state/relation.go`: Updated queries to use `relation_endpoint_uuid` instead of `relation_uuid`. Adjusted JOINs and conditions accordingly.
- `domain/relation/state/relation_test.go`: Updated test cases to align with the new data model, replacing `relation_uuid` with `relation_endpoint_uuid`.
- `domain/relation/state/types.go`: Added `RelationEndpointUUID` field in `relationUnit` and reflected changes in the data structure.
- `domain/schema/model/sql/0024-relation.sql`: Modified `relation_unit` table schema to replace `relation_uuid` with `relation_endpoint_uuid`. Adjusted constraints, indexes and views

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing

## QA steps

* Unit test should passes
* Deploy vault and easyrsa, integrate them. Should do something.

## Documentation changes



## Links

Suggested here: https://github.com/juju/juju/pull/19413#discussion_r2028425864

Aflynn50 commented on domain/relation/state/relation.go:
```diff
@@ -1272,9 +1279,14 @@ AND     unit_uuid = $relationUnit.unit_uuid
-VALUES ($relationUnit.*)
-ON CONFLICT (relation_uuid, unit_uuid) DO UPDATE SET
+INSERT INTO relation_unit (uuid, relation_endpoint_uuid, unit_uuid, in_scope) 
+SELECT $relationUnit.uuid, re.uuid, $relationUnit.unit_uuid, $relationUnit.in_scope
```
**question:** Does this query has to be so complicated. Can't you just insert it against the relation endpoint uuid you fetched further up, and not bother with the select?

hmlanigan replied:
@gfouillet todo: add a comment in the appropriate location for our future selves.

manadart replied:
This is simplified somewhat by my prior comment.

Aflynn50 reviewed (commented):
Looks good, though I think one of the queries is over-complicated.

hmlanigan commented on domain/relation/state/relation_test.go:
```diff
@@ -2483,11 +2489,11 @@ VALUES (?,?,?)
 
 // addRelationUnit inserts a relation unit into the database using the provided UUIDs for relation and unit.
-func (s *relationSuite) addRelationUnit(c *gc.C, unitUUID coreunit.UUID, relationUUID corerelation.UUID, relationUnitUUID corerelation.UnitUUID, inScope bool) {
+func (s *relationSuite) addRelationUnit(c *gc.C, unitUUID coreunit.UUID, relationEndpointUUID string, relationUnitUUID corerelation.UnitUUID, inScope bool) {
```
todo: Use the type please rather than a string.

gfouillet replied:
Agree, but won't do in this PR. EndpointUUID are always passed as string now in this test suite, mostly because they are not real uuid. 

I've the intent to refactor a bit this test to remove "fake-uuids" and builders, but it is a different piece of works.

hmlanigan commented on domain/relation/state/relation_test.go:
```diff
@@ -1556,11 +1556,17 @@ func (s *relationSuite) TestGetRelationUnit(c *gc.C) {
 	unitUUID := coreunittesting.GenUnitUUID(c)
 	relUUID := corerelationtesting.GenRelationUUID(c)
 	relUnitUUID := corerelationtesting.GenRelationUnitUUID(c)
+	relEndpointUUID := uuid.MustNewUUID().String()
+	applicationEndpointUUID := uuid.MustNewUUID().String()
+	charmRelationUUID := uuid.MustNewUUID().String()
```
todo: we've updated the tests to use the types rather than string versions to start, please update.

gfouillet replied:
Won't do: it will implies to rewrite the helper, which I want to in a further PR. This PR is scoped to ddl change, not the way test are written.

manadart commented on domain/schema/model/sql/0024-relation.sql:
```diff
@@ -84,20 +84,20 @@ ON relation (relation_id);
 -- The relation_unit table links a relation to a specific unit.
 CREATE TABLE relation_unit (
     uuid TEXT NOT NULL PRIMARY KEY,
```
It's the PK and has one implicitly.

manadart commented on domain/schema/model/sql/0024-relation.sql:
```diff
@@ -84,20 +84,20 @@ ON relation (relation_id);
-    relation_uuid TEXT NOT NULL,
+    relation_endpoint_uuid TEXT NOT NULL,
     unit_uuid TEXT NOT NULL,
     in_scope BOOLEAN DEFAULT FALSE,
```
We need to remove this `in_scope` field. A unit is in the relation scope by virtue of having a record in this table. 

So there is also no `upsertRelationUnitAndEnterScope` (above) either. It's one thing - inserting the relation unit.

Someone has gotten confused reading the old state code. When you call `Relation.Unit` there, it does not imply that here is any scope record for that unit in that relation, it creates a data structure, which as one of its functions can check if the unit is `InScope` by looking for it in the `relationscopes` collection.

manadart reviewed (changes requested):
The table change is correct, but we might as will fix the scope modelling while we're here.

gfouillet commented:
@manadart I removed scope, which made a bit more complicated the logic to get RelationUnitStatus while fetching all relation where the unit could be related.

I introduced a view to do that:

![GetRelationsStatusForUnit](https://github.com/user-attachments/assets/d7932d93-c1ff-4dcf-be14-2465e2f39f30)

manadart commented on domain/schema/model/sql/0024-relation.sql:
```diff
@@ -236,4 +233,16 @@ SELECT
-JOIN relation_status_type AS rst ON rs.relation_status_type_id = rst.id
+JOIN relation_status_type AS rst ON rs.relation_status_type_id = rst.id;
+
+CREATE VIEW v_relation_unit_status AS
```
Given that the use of this view is quite esoteric - a unit-eye view of all relations its application is in, it's unlikely to be useful outside the single usage in state.

Just put this query directly where the view is being selected from.

manadart reviewed (approved):
Thanks for the changes. Drop the view and we're good to go.

gfouillet commented:
/merge
